### PR TITLE
Synchronise player tools and focus between clients

### PR DESCRIPTION
**Pull request by CWMlolzlz**
Based on the feature request https://github.com/CitiesSkylinesMultiplayer/CSM/issues/198

# Overview
This feature aims to synchronize each clients tool state with other player so players can see actions other players are performing. From some early attempts it is possible to share the internal state of each player's active tool and hook it into render loop of the game. There is no need to reimplement the render logic for each tool and only the state required for rendering each tool needs to be shared.

# Features
- [x] Coloring each player's rendered tools
- [x] Sync default/idle tool
- [x] Sync net tool
- [x] Sync building tool
- [x] Sync prop tool
- [x] Sync zoning tool
- [x] Sync bulldoze tool
- [x] Sync tree tool
- [x] Sync terrain tool
- [x] Sync district tool
- [ ] Sync disaster tool
- [ ] Sync transport tool
- [ ] Sync resource tool? ^
- [ ] Sync water tool? ^
- [ ] Sync surface tool? ^
- [ ] Sync snapshot tool? ^
- [ ] Sync game area tool? ^^

^ These tools aren't normally available within the normal game. We would want to support it so if a mod exposes it in gameplay it will be synced.
^^ Doesn't have dedicated rendering logic.

# Future Ideas
- Render other player cursors over their current position. This helps convey information that overlay rendering doesn't reveal. Super important for zoning tool and terrain tool.
- Drawing tool. Allow users to draw onto terrain temporarily for communication. Achievable with `LineRenderer`
- Pinging. Lets users ping a location to notify others

# Bugs
- [x] NetTool overlays from other clients are rendered if they too are in the NetTool

# Technical notes
## `ToolSimulator`
The `ToolSimulator` is the critical component that is responsible for holding onto the synced state of each tool. I have made this implement the `Singleton<T>` class so it exists as a `GameObject` in the Unity engine. Very useful for debugging with ModTools.

## Maintenance
Since this mod is reading and writing from private fields within the game's tools there is a risk that a change in a field name can break the syncing of tools. it would likely be on a tool-by-tool basis.

### CODE REVIEW

**Comment by CWMlolzlz** [comment]
@kaenganxt I accidentally created this without setting it to draft. Would you be able to change this to a draft, I don't think I have permission

**Comment by kaenganxt** [comment]
What's the state on this? Are only the three remaining tools missing before this is ready?

**Comment by RuisSoftware** [comment]
Please continue working on this pull request :)

**Comment by kaenganxt** [comment]
I changed/implemented the following things in my new commits:
- Move tool handlers to basegame project
- Implement the district tool
- Fix some default cursors
- Remove old player pointer logic
- Make tool commands non-transactional
- Remove some debug code
- Add some null checks to UI logic
- Don't use the color red for other players
- Display brushes correctly
- Fix packet equality checks to prevent spamming
- Fix net tool showing angles forever
- Destroy old cursor objects to prevent exceptions on unload

The transport tool implementation is still not working correctly, but this is caused by the basic transport line synchronisation being bugged when two players build lines at the same time.

The disaster tool will be looked at when implementing DLC features.
So I think this PR is ready for merging at this point.

The PR should build correctly once the Cities assemblies are up to date.

**Comment by DominicMaas** [comment]
I'll update the assemblies when I get home this afternoon

**Comment by DominicMaas** [comment]
Updated the assemblies, lets see how this goes

File: src/api/Networking/Player.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿using CSM.API.Networking.Status;
+﻿using System;
```
Comment (hibo98) [inline src/api/Networking/Player.cs]: unused import

File: src/basegame/Injections/Tools/DistrictToolHandler.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            }
+			if (tool.m_mode == DistrictTool.Mode.Paint)
+			{
+				return existing.m_airportAreaCursor;
+			}
+			else
+			{
+				return existing.m_eraseCursor;
+			}
+        }
```
Comment (hibo98) [inline src/basegame/Injections/Tools/DistrictToolHandler.cs]: Check for using spaces instead of tabs

File: src/csm/CSM.cs [hunk 3]
```diff
@@ -2,10 +2,12 @@
 using CitiesHarmony.API;
 using ColossalFramework.IO;
 using CSM.API;
+using CSM.API.Networking;
 using CSM.Commands;
 using CSM.GS.Commands;
 using CSM.Injections;
 using CSM.Mods;
+using CSM.Networking;
```
Comment (hibo98) [inline src/csm/CSM.cs]: unused imports

File: src/csm/Networking/MultiplayerManager.cs [hunk 4]
```diff
@@ -5,6 +5,7 @@
 using ColossalFramework.UI;
 using CSM.API;
 using CSM.API.Commands;
+using CSM.API.Networking;
```
Comment (hibo98) [inline src/csm/Networking/MultiplayerManager.cs]: unused import

**Review by hibo98 — CHANGES_REQUESTED** [review]

**Review by hibo98 — APPROVED** [review]

**Pull request merged**